crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -174,11 +178,42 @@ impl ExpectedDocstringKind {
     }
 }
 
+struct ParsedAnnotationsCache<'a> {
+    arena: &'a typed_arena::Arena<AnnotationParseResult>,
```

Comment:
Nit: I'm leaning towards caching only successfully parsed annotations, considering that annotations with syntax errors should be rare. This would help reduce the size of every element in the arena.